Data/Aeson/Types/Class.hs

```diff
@@ -272,3 +320,95 @@ typeMismatch expected actual =
+    liftToEncodingList f g = list (liftToEncoding f g)
+
+-- | Lift the standard 'toJSON' function through the type constructor.
+toJSON1 :: (ToJSON1 f, ToJSON a) => f a -> Value
+toJSON1 = liftToJSON toJSON toJSONList
+{-# INLINE toJSON1 #-}
+
+-- | Lift the standard 'toEncoding' function through the type constructor.
+toEncoding1 :: (ToJSON1 f, ToJSON a) => f a -> Encoding
+toEncoding1 = liftToEncoding toEncoding toEncodingList
+{-# INLINE toEncoding1 #-}
+
+
+-- | Lifting of the 'FromJSON' class to binary type constructors.
+class FromJSON2 f where
+    liftParseJSON2
+        :: (Value -> Parser a)
+        -> (Value -> Parser [a])
+        -> (Value -> Parser b)
+        -> (Value -> Parser [b])
+        -> Value -> Parser (f a b)
+    liftParseJSONList2 
+        :: (Value -> Parser a) 
+        -> (Value -> Parser [a]) 
+        -> (Value -> Parser b) 
+        -> (Value -> Parser [b]) 
+        -> Value -> Parser [f a b]
+    liftParseJSONList2 fa ga fb gb v = case v of
+        Array vals -> fmap V.toList (V.mapM (liftParseJSON2 fa ga fb gb) vals)
+        _ -> typeMismatch "[a]" v
+
+-- | Lift the standard 'parseJSON' function through the type constructor.
+parseJSON2 :: (FromJSON2 f, FromJSON a, FromJSON b) => Value -> Parser (f a b)
+parseJSON2 = liftParseJSON2 parseJSON parseJSONList parseJSON parseJSONList
+{-# INLINE parseJSON2 #-}
+
+-- | Lifting of the 'ToJSON' class to binary type constructors.
+class ToJSON2 f where
+    liftToJSON2 :: (a -> Value) -> ([a] -> Value) -> (b -> Value) -> ([b] -> Value) -> f a b -> Value
+    liftToJSONList2 :: (a -> Value) -> ([a] -> Value) -> (b -> Value) -> ([b] -> Value) -> [f a b] -> Value
+    liftToJSONList2 fa ga fb gb = listValue (liftToJSON2 fa ga fb gb)
+
+    liftToEncoding2 :: (a -> Encoding) -> ([a] -> Encoding) -> (b -> Encoding) -> ([b] -> Encoding) -> f a b -> Encoding
+    liftToEncodingList2 :: (a -> Encoding) -> ([a] -> Encoding) -> (b -> Encoding) -> ([b] -> Encoding) -> [f a b] -> Encoding
+    liftToEncodingList2 fa ga fb gb = list (liftToEncoding2 fa ga fb gb)
+
+-- | Lift the standard 'toJSON' function through the type constructor.
+toJSON2 :: (ToJSON2 f, ToJSON a, ToJSON b) => f a b -> Value
+toJSON2 = liftToJSON2 toJSON toJSONList toJSON toJSONList
+{-# INLINE toJSON2 #-}
+
+-- | Lift the standard 'toEncoding' function through the type constructor.
+toEncoding2 :: (ToJSON2 f, ToJSON a, ToJSON b) => f a b -> Encoding
+toEncoding2 = liftToEncoding2 toEncoding toEncodingList toEncoding toEncodingList
+{-# INLINE toEncoding2 #-}
+
+list :: (a -> Encoding) -> [a] -> Encoding
+list _  []     = emptyArray_
+list to (x:xs) = Encoding $
+                B.char7 '[' <> fromEncoding (to x) <> commas xs <> B.char7 ']'
+      where commas = foldr (\v vs -> B.char7 ',' <> fromEncoding (to v) <> vs) mempty
+{-# INLINE list #-}
+
+listValue :: (a -> Value) -> [a] -> Value
+listValue f = Array . V.fromList . map f
```

Comment:
I agree. I had only left it named `list` here because I stole it out of `Data.Aeson.Encode.Functions`, but `listEncoding` is a much better name.